spec/policy/tls_validation/tls_validation_spec.lua

```diff
@@ -0,0 +1,75 @@
+local _M = require('apicast.policy.tls_validation')
+
+local server = assert(fixture('CA', 'server.crt'))
+local CA = assert(fixture('CA', 'CA.crt'))
+local client = assert(fixture('CA', 'client.crt'))
+
+describe('tls_validation policy', function()
+  describe('.new', function()
+    it('works without configuration', function()
+      assert(_M.new())
+    end)
+
+    it('accepts configuration', function()
+      assert(_M.new({
+        whitelist = {
+          { pem_certificate = [[--BEGIN CERTIFICATE--]] }
+        }
+      }))
+    end)
+  end)
+
+  describe(':access', function()
+    before_each(function()
+      stub(ngx, 'say')
+      stub(ngx, 'exit')
+    end)
+
+    it('rejects non whitelisted certificate', function()
+      ngx.var = { ssl_client_raw_cert = client }
+
+      local policy = _M.new({ whitelist = { { pem_certificate = server }}})
+
+      policy:access()
+
+      assert.stub(ngx.exit).was_called_with(400)
```

Comment:
Would 403 be more appropriate?

Reply:
I think I got the 400 from testing normal nginx certificate veritifation.